roles/jd-server/src/lib/job_declarator/message_handler.rs

```diff
@@ -154,33 +154,48 @@ impl ParseClientJobDeclarationMessages for JobDeclaratorDownstream {
         &mut self,
         message: ProvideMissingTransactionsSuccess,
     ) -> Result<SendTo, Error> {
-        let (_, ref mut transactions_with_state, missing_indexes) = &mut self.declared_mining_job;
+        let (declared_mining_job, ref mut transactions_with_state, missing_indexes) =
+            &mut self.declared_mining_job;
         let mut unknown_transactions: Vec<Transaction> = vec![];
-        for (i, tx) in message.transaction_list.inner_as_ref().iter().enumerate() {
-            let mut cursor = Cursor::new(tx);
-            let transaction = Transaction::consensus_decode_from_finite_reader(&mut cursor)
-                .map_err(|e| Error::TxDecodingError(e.to_string()))?;
-            Vec::push(&mut unknown_transactions, transaction.clone());
-            let index = *missing_indexes
-                .get(i)
-                .ok_or(Error::LogicErrorMessage(Box::new(
-                    AllMessages::JobDeclaration(JobDeclaration::ProvideMissingTransactionsSuccess(
-                        message.clone().into_static(),
-                    )),
-                )))? as usize;
-            // insert the missing transactions in the mempool
-            transactions_with_state[index] = TransactionState::PresentInMempool(transaction.txid());
-        }
-        self.add_txs_to_mempool
-            .add_txs_to_mempool_inner
-            .unknown_transactions
-            .append(&mut unknown_transactions);
-        // if there still a missing transaction return an error
-        for tx_with_state in transactions_with_state {
-            match tx_with_state {
-                TransactionState::PresentInMempool(_) => continue,
-                TransactionState::Missing => return Err(Error::JDSMissingTransactions),
+        match declared_mining_job {
+            Some(declared_job) => {
+                let id = declared_job.request_id;
+                // check request_id in order to ignore old ProvideMissingTransactionsSuccess (see issue #860)
+                if id == message.request_id {
```

Comment:
I think we're still not fully fixing the issue

regardless of this new conditional on line 164, execution continues after line 200, and JDS still returns `DeclareMiningJobSuccess`, even if `request_id` is wrong

Reply:
this could be even seen as a potential attack vector:

JDS sent a `ProvideMissingTransaction` as a verification that `declared_job` is valid

but JDC could send any garbage on `ProvideMissingTransaction.Success` and still get a `DeclareMiningJobSuccess` in return

Reply:
Do you think it would be better to send a DeclareMiningJobError (with a new error code like "declared-mining-job-old"?
I implemented this solution after discussion and suggestion by @Fi3 here: https://github.com/stratum-mining/stratum/issues/860#issuecomment-2095386031

Reply:
Returning an error is wrong. The client is sending a message that in this context is valid. Either you keep track of older declared job and you handle this message using the old job (pointless imo since we are going to use a newer job) or you just ignore the message, you don't have to answer with something.

Reply:
we want to avoid sending `DeclareMiningJobSuccess` if `id != message.request_id`

so I think we need to move the logic between lines 200-211 inside the new conditional on line 164 (`if id == message.request_id`)

and as @Fi3 pointed out, leave the generic case (last line of the function implementation) as `Ok(SendTo::None(None))`

Reply:
Done in 0ef37d6ad216bebb2b00fbec0d6ad792355ffef3

Reply:
> Returning an error is wrong. The client is sending a message that in this context is valid. Either you keep track of older declared job and you handle this message using the old job (pointless imo since we are going to use a newer job) or you just ignore the message, you don't have to answer with something. 
> 
> `SendTo::None(None)`

I disagree 

it doesn't matter that `ProvideMissingTransaction.Success` is correct 

the `DeclareMiningJob.Error` is not a response to this message, it is a signal that JDS will not accept that job because it is no longer valid

if we bypass this, JDS is giving a false acknowledgement, and when the shares are rejected in the future (which will happen), JDC will trigger a fallback unnecessarily

Reply:
> > Returning an error is wrong. The client is sending a message that in this context is valid. Either you keep track of older declared job and you handle this message using the old job (pointless imo since we are going to use a newer job) or you just ignore the message, you don't have to answer with something.
> > `SendTo::None(None)`
> 
> I disagree
> 
> it doesn't matter that `ProvideMissingTransaction.Success` is correct in this context
> 
> the `DeclareMiningJob.Error` is not a response to this specific message, it is a signal on a broader context, that JDS will not accept that job because it is no longer valid
> 
> if we bypass this, JDS is giving a false acknowledgement, and when the shares are rejected in the future (which will happen), JDC will trigger a fallback unnecessarily

The point here is that you are not sending DeclareMiningJob.Success so the no one is saying that a job is accepeted

Reply:
I have to double check it but in principle I am against changing the spec, specially if we can handle it without changing the spec.

Reply:
@GitGab19
I do not get how this error can happen. Do you know the exact series of message that lead to it? I tried to figure out, it but I am at an endpoint.

So let's say that we have:

```
######case 1:######
------template1------------
new_template1 TP -> JDC
snp1 TP -> JDC
req_tx_data1 JDC -> TP
tx_data1 TP -> JDC
declare1 JDC -> JDS
miss_data1 JDS -> JDC
------template2------------
new_template2 TP -> JDC
snp2 TP -> JDC
req_tx_data2 JDC -> TP
tx_data2 TP -> JDC
declare2 JDC -> JDS
miss_data2 JDS -> JDC
prov_miss_tx1 JDC -> JDS
prov_miss_tx2 JDC -> JDS
declate_success_2 -> JDS -> JDC
```

In this case we are going to spawn 2 task one for snp1 and the other for snp2.
here snp2 can get the declared2 job and exit from the loop, snp1 will never exit but we could put a timeout.

So from what I see here I don't see any issue related with the JDS not sending the success message. A part from the obvious fact the we need a timeout for the on_set_new_prev_hash_task

Reply:
This is the model that I use to reason about it:

```
# JDC (TP module)

on_new_template:
 * fetch txs data from template, it will answear with req tx data success
 * self.new_template_message = new_template

on_set_new_p_hash:
 * JobDeclarator::on_set_new_prev_hash(..)

on_req_tx_data_succ:
 * get new_template from self.new_template_message
 * call JobDeclarator::on_new_template(..) with new_tamplate

# JDC (jd module)

on_new_template:
 * self.last_declare_job_sent = Some(declare_job)
 * send declare mining job to jds 

on_set_new_p_hash:
 * itereate over self.future_jobs.remove(id), until it get the associated job and template

on receiving declare mining job success from upstream:
  if future:
    insert lastb declared job into self.future_jobs
```

Reply:
on_req_tx_data_succ:
 * get new_template from self.new_template_message
 * call JobDeclarator::on_new_template(..) with new_tamplate
 
 
 Here we don't use ids we could have a bug here maybe

Reply:
yep I think that something like what we have above could be the cause of this issue

Reply:
> The point here is that you are not sending DeclareMiningJob.Success so the no one is saying that a job is accepeted

I think we missed a detail on this thread, but @GitGab19 told me on Discord that JDC needs to receive **some** message in order to be able to exit a loop. So having JDS ignore this scenario without sending anything is not an option.

This message needs to be either `DeclareMiningMessage.Success` or `DeclareMiningMessage.Error` (or maybe `IdentifyTransactions`, but we don't have any logic for this message so we can ignore this case).

As stated previously, sending `DeclareMiningMessage.Success` is not desirable, because it would be a misleading acknowledgement of a job that will produce shares that end up being rejected.

What I disagreed on my previous comment is that we shouldn't `DeclareMiningMessage.Error`. It doesn't matter that `ProvideMissingTransaction.Success` is correct. The job on this `request_id` is no longer valid so JDC needs to know that, and `DeclareMiningMessage.Error` is the only way it will know that.

With regards to error code (and changing the spec), adding a `old-request-id` seems reasonable.

Reply:
Can you point out why not receiving a message is not an option?

Reply:
I missed the suggestion on the timeout, sorry about that. Maybe that's a valid option too.

I'm only strongly opposed to returning a `DeclareMiningJob.Success` if the job is no longer valid, but I guess we all agree that's not a good solution.

So whatever you guys decide here is fine for me.

Reply:
I just managed to exit from the loop in JDC without the necessity of receiving a DeclareMiningJob.Success/Error, and without arbitrary timeouts.
I used a counter to check the number of tasks created into `on_new_prev_hash` function, added here: 0f283109f29e735764e2123960f431e91dca4f49.

It means that this PR is ready to be merged, since no changes are needed from last review made by @plebhash.

Reply:
cool, it needs to be rebased tho, and it seems I cannot do it on Github UI:
![image](https://github.com/stratum-mining/stratum/assets/147345153/eac836dd-cb80-43d1-b4e9-69969b5ed6b7)

Reply:
Just rebased 

> cool, it needs to be rebased tho, and it seems I cannot do it on Github UI: ![image](https://private-user-images.githubusercontent.com/147345153/331372202-eac836dd-cb80-43d1-b4e9-69969b5ed6b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4OTUyMTgsIm5iZiI6MTcxNTg5NDkxOCwicGF0aCI6Ii8xNDczNDUxNTMvMzMxMzcyMjAyLWVhYzgzNmRkLWNiODAtNDNkMS1iNGU5LTY5OTY5YjVlZDZiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNlQyMTI4MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01Y2MxYjMxYWJmNzgxNGZlMzA3ZDJhMzA2ZDFlZjQzZDA1M2RjNmE2ODMwMGJjMDMzMzgwMTJjOTNmZTc4OTgzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.o2ypYvrk8WsfEL5FFFY1NbB9Xm9D2113gg9UCSdjneU)

Just rebased and pushed